### ISSUE TITLE
Bouncer not deleting ipsets

### ISSUE THREAD

**Issue by lgg42** [crowdsecurity/cs-aws-waf-bouncer#72]
Hello everyone! I'm running a Crowdsec installation for 3 services supposedly fine (I get IP bans in the correct scenarios) until I received an error in one of the bouncer logs stating that it couldn't create more new AWS WAF IPSets. I realized I had 100 existing IPSets and that was a current limit that I'd need to increase.

I have 3 EC2 instances. Each instance runs a different service via docker-compose stack. And in each stack there's a crowdsec and crowdsec-awf-waf-bouncer service running.

All three services share the same AWS WAF ACL (crowdsec-<ENV_NAME>) and each service writes a new Group Rule. Here's the example configuration for the bouncer of the service "myservice":

```yml
api_key: redacted-api-key
api_url: "http://127.0.0.1:8080/"
update_frequency: 10s
waf_config:
  - web_acl_name: crowdsec-staging
    fallback_action: ban
    rule_group_name: crowdsec-waf-bouncer-ip-set-myservice
    scope: REGIONAL
    capacity: 300
    region: us-east-1
    ipset_prefix: myservice-crowdsec-ipset-a
```

From https://docs.crowdsec.net/u/bouncers/aws_waf/ for the `ipset_prefix` parameter it states: **"All ipsets are deleted on shutdown."**

And I noticed this is not happening. Everytime the docker-compose stack is restarted new IPSets are created and the old ones remain.

I have RTFM and STFW without results. I have no suspicious information from the logs of **crowdsec** and **crowdsec-awf-waf-bouncer** that I can use.

I have tried setting IAM AdministratorAccess policy to the EC2's IAM role in case it was lacking an IAM permissions but it seems not to be the case.

Has anyone detected this issue before? What could I be doing wrong?

Thanks in advance for reading.

### Versions
Crowdsec: crowdsecurity/crowdsec:v1.6.2
Bouncer: crowdsecurity/aws-waf-bouncer:v0.1.7

**Comment by blotus** [comment]
Hello,

We are using this bouncer internally, and this is the first time we have seen this error.
I've just tried to stop the bouncer, and the resources are properly deleted from the account:
```
time="2024-10-31T13:53:25Z" level=info msg="terminating bouncer process"
time="2024-10-31T13:53:25Z" level=info msg="Cleaning up resources" acl=cti_public_api_searchbar-rule region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:26Z" level=info msg="deleting set crowdsec-ipset--IPV4-ban-33664642-5547-420d-b7fb-5ae02f5e714d" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:26Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV4-ban-33664642-5547-420d-b7fb-5ae02f5e714d" set=crowdsec-ipset--IPV4-ban-33664642-5547-420d-b7fb-5ae02f5e714d
time="2024-10-31T13:53:27Z" level=info msg="deleting set crowdsec-ipset--IPV4-ban-0c57b313-b6d7-44de-bbaf-f717dba6dc63" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:27Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV4-ban-0c57b313-b6d7-44de-bbaf-f717dba6dc63" set=crowdsec-ipset--IPV4-ban-0c57b313-b6d7-44de-bbaf-f717dba6dc63
time="2024-10-31T13:53:30Z" level=info msg="deleting set crowdsec-ipset--IPV4-ban-f5477bf8-d287-40b5-afed-67565b64c0d3" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:30Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV4-ban-f5477bf8-d287-40b5-afed-67565b64c0d3" set=crowdsec-ipset--IPV4-ban-f5477bf8-d287-40b5-afed-67565b64c0d3
time="2024-10-31T13:53:31Z" level=info msg="deleting set crowdsec-ipset--IPV6-ban-07416271-3ddb-4bf1-9764-796ed47c34c4" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:31Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV6-ban-07416271-3ddb-4bf1-9764-796ed47c34c4" set=crowdsec-ipset--IPV6-ban-07416271-3ddb-4bf1-9764-796ed47c34c4
time="2024-10-31T13:53:32Z" level=info msg="Cleaning up resources" acl=hub-cdn-acl region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:33Z" level=info msg="deleting set hub-cdn-ipset--IPV4-ban-aa688add-0a1b-47dc-b471-85228c28aba8" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:33Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV4-ban-aa688add-0a1b-47dc-b471-85228c28aba8" set=hub-cdn-ipset--IPV4-ban-aa688add-0a1b-47dc-b471-85228c28aba8
time="2024-10-31T13:53:35Z" level=info msg="deleting set hub-cdn-ipset--IPV4-ban-02c3b350-0ab2-4d22-9694-67c29e5d3138" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:35Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV4-ban-02c3b350-0ab2-4d22-9694-67c29e5d3138" set=hub-cdn-ipset--IPV4-ban-02c3b350-0ab2-4d22-9694-67c29e5d3138
time="2024-10-31T13:53:36Z" level=info msg="deleting set hub-cdn-ipset--IPV4-ban-e1ab12ec-f642-42e0-8803-39ceaf80d23d" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:36Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV4-ban-e1ab12ec-f642-42e0-8803-39ceaf80d23d" set=hub-cdn-ipset--IPV4-ban-e1ab12ec-f642-42e0-8803-39ceaf80d23d
time="2024-10-31T13:53:37Z" level=info msg="deleting set hub-cdn-ipset--IPV6-ban-5aebefc2-82cf-4c71-90cf-21be0907a35b" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
time="2024-10-31T13:53:37Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV6-ban-5aebefc2-82cf-4c71-90cf-21be0907a35b" set=hub-cdn-ipset--IPV6-ban-5aebefc2-82cf-4c71-90cf-21be0907a35b
```

For reference, we have the following policy attached to the instance (which is admittedly a bit too large if you have other ACLs you don't want the bouncer to have access to):
```
 {
            "Action": [
                "wafv2:UpdateWebACL",
                "wafv2:UpdateRuleGroup",
                "wafv2:UpdateIPSet",
                "wafv2:TagResource",
                "wafv2:ListWebACLs",
                "wafv2:ListRuleGroups",
                "wafv2:ListIPSets",
                "wafv2:GetWebACL",
                "wafv2:GetRuleGroup",
                "wafv2:GetIPSet",
                "wafv2:DeleteRuleGroup",
                "wafv2:DeleteIPSet",
                "wafv2:CreateRuleGroup",
                "wafv2:CreateIPSet"
            ],
            "Effect": "Allow",
            "Resource": [
                "arn:aws:wafv2:us-east-1:XXXX:global/webacl/XXXX/XXXXX",
                "arn:aws:wafv2:us-east-1:XXXX:global/webacl/*/*",
                "arn:aws:wafv2:us-east-1:XXXX:global/rulegroup/*/*",
                "arn:aws:wafv2:us-east-1:XXXX:global/managedruleset/*/*",
                "arn:aws:wafv2:us-east-1:XXXX:*/ipset/*/*"
            ],
            "Sid": "ManageWAF"
        },
```
        
        
 Do you see the `Deleting IPSet ....` in your logs when stopping the bouncer ?
 How many ACLs is the bouncer configured to manage ? If you have a lot, there might be an issue with the rate limiting on AWS side.

**Comment by lgg42** [comment]
> Hello,
> 
> We are using this bouncer internally, and this is the first time we have seen this error. I've just tried to stop the bouncer, and the resources are properly deleted from the account:
> 
> ```
> time="2024-10-31T13:53:25Z" level=info msg="terminating bouncer process"
> time="2024-10-31T13:53:25Z" level=info msg="Cleaning up resources" acl=cti_public_api_searchbar-rule region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:26Z" level=info msg="deleting set crowdsec-ipset--IPV4-ban-33664642-5547-420d-b7fb-5ae02f5e714d" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:26Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV4-ban-33664642-5547-420d-b7fb-5ae02f5e714d" set=crowdsec-ipset--IPV4-ban-33664642-5547-420d-b7fb-5ae02f5e714d
> time="2024-10-31T13:53:27Z" level=info msg="deleting set crowdsec-ipset--IPV4-ban-0c57b313-b6d7-44de-bbaf-f717dba6dc63" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:27Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV4-ban-0c57b313-b6d7-44de-bbaf-f717dba6dc63" set=crowdsec-ipset--IPV4-ban-0c57b313-b6d7-44de-bbaf-f717dba6dc63
> time="2024-10-31T13:53:30Z" level=info msg="deleting set crowdsec-ipset--IPV4-ban-f5477bf8-d287-40b5-afed-67565b64c0d3" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:30Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV4-ban-f5477bf8-d287-40b5-afed-67565b64c0d3" set=crowdsec-ipset--IPV4-ban-f5477bf8-d287-40b5-afed-67565b64c0d3
> time="2024-10-31T13:53:31Z" level=info msg="deleting set crowdsec-ipset--IPV6-ban-07416271-3ddb-4bf1-9764-796ed47c34c4" acl=cti_public_api_searchbar-rule component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:31Z" level=info msg="Deleting IPSet crowdsec-ipset--IPV6-ban-07416271-3ddb-4bf1-9764-796ed47c34c4" set=crowdsec-ipset--IPV6-ban-07416271-3ddb-4bf1-9764-796ed47c34c4
> time="2024-10-31T13:53:32Z" level=info msg="Cleaning up resources" acl=hub-cdn-acl region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:33Z" level=info msg="deleting set hub-cdn-ipset--IPV4-ban-aa688add-0a1b-47dc-b471-85228c28aba8" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:33Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV4-ban-aa688add-0a1b-47dc-b471-85228c28aba8" set=hub-cdn-ipset--IPV4-ban-aa688add-0a1b-47dc-b471-85228c28aba8
> time="2024-10-31T13:53:35Z" level=info msg="deleting set hub-cdn-ipset--IPV4-ban-02c3b350-0ab2-4d22-9694-67c29e5d3138" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:35Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV4-ban-02c3b350-0ab2-4d22-9694-67c29e5d3138" set=hub-cdn-ipset--IPV4-ban-02c3b350-0ab2-4d22-9694-67c29e5d3138
> time="2024-10-31T13:53:36Z" level=info msg="deleting set hub-cdn-ipset--IPV4-ban-e1ab12ec-f642-42e0-8803-39ceaf80d23d" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:36Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV4-ban-e1ab12ec-f642-42e0-8803-39ceaf80d23d" set=hub-cdn-ipset--IPV4-ban-e1ab12ec-f642-42e0-8803-39ceaf80d23d
> time="2024-10-31T13:53:37Z" level=info msg="deleting set hub-cdn-ipset--IPV6-ban-5aebefc2-82cf-4c71-90cf-21be0907a35b" acl=hub-cdn-acl component=ipset_manager region=us-east-1 scope=CLOUDFRONT
> time="2024-10-31T13:53:37Z" level=info msg="Deleting IPSet hub-cdn-ipset--IPV6-ban-5aebefc2-82cf-4c71-90cf-21be0907a35b" set=hub-cdn-ipset--IPV6-ban-5aebefc2-82cf-4c71-90cf-21be0907a35b
> ```
> 
> For reference, we have the following policy attached to the instance (which is admittedly a bit too large if you have other ACLs you don't want the bouncer to have access to):
> 
> ```
>  {
>             "Action": [
>                 "wafv2:UpdateWebACL",
>                 "wafv2:UpdateRuleGroup",
>                 "wafv2:UpdateIPSet",
>                 "wafv2:TagResource",
>                 "wafv2:ListWebACLs",
>                 "wafv2:ListRuleGroups",
>                 "wafv2:ListIPSets",
>                 "wafv2:GetWebACL",
>                 "wafv2:GetRuleGroup",
>                 "wafv2:GetIPSet",
>                 "wafv2:DeleteRuleGroup",
>                 "wafv2:DeleteIPSet",
>                 "wafv2:CreateRuleGroup",
>                 "wafv2:CreateIPSet"
>             ],
>             "Effect": "Allow",
>             "Resource": [
>                 "arn:aws:wafv2:us-east-1:XXXX:global/webacl/XXXX/XXXXX",
>                 "arn:aws:wafv2:us-east-1:XXXX:global/webacl/*/*",
>                 "arn:aws:wafv2:us-east-1:XXXX:global/rulegroup/*/*",
>                 "arn:aws:wafv2:us-east-1:XXXX:global/managedruleset/*/*",
>                 "arn:aws:wafv2:us-east-1:XXXX:*/ipset/*/*"
>             ],
>             "Sid": "ManageWAF"
>         },
> ```
> 
> Do you see the `Deleting IPSet ....` in your logs when stopping the bouncer ? How many ACLs is the bouncer configured to manage ? If you have a lot, there might be an issue with the rate limiting on AWS side.

Thank you for taking a time to look into this issue :pray: 

I set the `log_level: debug` in /crowdsec-aws-waf-bouncer.yaml and proceeded to check your questions.

First of all, I do not see the `Deleting IPSet ....` lines in the container. When the `docker-compose stop crowdsec-awf-waf-bouncer` is executed this are all the logs the container stdout's:

```
time="2024-11-03T16:48:18Z" level=debug msg="[URL] GET http://127.0.0.1:8080/v1/decisions/stream?scopes=ip%2Crange%2Ccountry"
time="2024-11-03T16:48:18Z" level=debug msg="req-api: GET http://127.0.0.1:8080/v1/decisions/stream?scopes=ip%2Crange%2Ccountry"
time="2024-11-03T16:48:18Z" level=debug msg="resp-api: http 200"
time="2024-11-03T16:48:18Z" level=debug msg="[headers] Date: [Sun, 03 Nov 2024 16:48:18 GMT]"
time="2024-11-03T16:48:18Z" level=debug msg="[headers] Content-Length: [27]"
time="2024-11-03T16:48:18Z" level=debug msg="[headers] Content-Type: [application/json; charset=utf-8]"
time="2024-11-03T16:48:18Z" level=debug msg="Response: HTTP/1.1 200 OK\r\nContent-Length: 27\r\nContent-Type: application/json; charset=utf-8\r\nDate: Sun, 03 Nov 2024 16:48:18 GMT\r\n\r\n{\"deleted\":null,\"new\":null}"
time="2024-11-03T16:48:18Z" level=info msg="Polling decisions"
time="2024-11-03T16:48:19Z" level=debug msg="checking if set myapp-crowdsec-ipset-a-IPV4-ban-27e956d1-96fc-4184-86ed-abdf2e84e9a4 is stale" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="checking if set myapp-crowdsec-ipset-a-IPV4-ban-66c4a49c-f129-4bc6-bee4-1bbf0c9ec732 is stale" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="checking if set myapp-crowdsec-ipset-a-IPV4-ban-51912a1d-6b44-4949-923e-cdc02a296dc7 is stale" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="checking if set myapp-crowdsec-ipset-a-IPV4-ban-f2657a6b-e946-46af-a64f-915abf584673 is stale" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="checking if set myapp-crowdsec-ipset-a-IPV4-ban-acc8e690-5c1a-49b4-a70e-f2e4e85c8b9f is stale" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="checking if set myapp-crowdsec-ipset-a-IPV6-ban-3cc5f962-fdb9-47b7-be93-5840851fc689 is stale" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="Multiple ipsets for action ban" acl=crowdsec-staging region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="No ipset for action captcha" acl=crowdsec-staging region=us-east-1 scope=REGIONAL
time="2024-11-03T16:48:19Z" level=debug msg="No ipset for action count" acl=crowdsec-staging region=us-east-1 scope=REGIONAL
```

for more information, and back to `log_level: info`, when I start the bouncer container (`docker-compose start crowdsec-awf-waf-bouncer`) this are the logs I get:

```
time="2024-11-03T16:49:37Z" level=info msg="Starting crowdsec-aws-waf-bouncer v0.1.7-ae18a8d09d6186117d570b229d698979edd50b7e"
time="2024-11-03T16:49:37Z" level=info msg="Using API key auth"
time="2024-11-03T16:49:38Z" level=info msg="Adding RuleGroup crowdsec-waf-bouncer-ip-set-myapp to ACL crowdsec-staging" acl=crowdsec-staging region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:39Z" level=info msg="Starting processing decisions"
time="2024-11-03T16:49:41Z" level=info msg="Polling decisions"
time="2024-11-03T16:49:41Z" level=info msg="could not find empty set, creating new set" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:41Z" level=info msg="could not find empty set, creating new set" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:43Z" level=info msg="could not find empty set, creating new set" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:45Z" level=info msg="could not find empty set, creating new set" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:48Z" level=info msg="could not find empty set, creating new set" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:48Z" level=info msg="could not find empty set, creating new set" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:48Z" level=info msg="set myapp-crowdsec-ipset-a-IPV4-ban-d4fd1200-6d15-48c6-a4f1-c3e3a408c211 is stale, updating it" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:48Z" level=info msg="Updating IPSet myapp-crowdsec-ipset-a-IPV4-ban-d4fd1200-6d15-48c6-a4f1-c3e3a408c211" set=myapp-crowdsec-ipset-a-IPV4-ban-d4fd1200-6d15-48c6-a4f1-c3e3a408c211
time="2024-11-03T16:49:48Z" level=info msg="Creating IPSet myapp-crowdsec-ipset-a-IPV4-ban-d4fd1200-6d15-48c6-a4f1-c3e3a408c211" set=myapp-crowdsec-ipset-a-IPV4-ban-d4fd1200-6d15-48c6-a4f1-c3e3a408c211
time="2024-11-03T16:49:49Z" level=info msg="set myapp-crowdsec-ipset-a-IPV4-ban-60386d43-c46a-451d-8714-bef7cd6bd5f5 is stale, updating it" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:49Z" level=info msg="Updating IPSet myapp-crowdsec-ipset-a-IPV4-ban-60386d43-c46a-451d-8714-bef7cd6bd5f5" set=myapp-crowdsec-ipset-a-IPV4-ban-60386d43-c46a-451d-8714-bef7cd6bd5f5
time="2024-11-03T16:49:49Z" level=info msg="Creating IPSet myapp-crowdsec-ipset-a-IPV4-ban-60386d43-c46a-451d-8714-bef7cd6bd5f5" set=myapp-crowdsec-ipset-a-IPV4-ban-60386d43-c46a-451d-8714-bef7cd6bd5f5
time="2024-11-03T16:49:49Z" level=info msg="Polling decisions"
time="2024-11-03T16:49:53Z" level=info msg="set myapp-crowdsec-ipset-a-IPV4-ban-01ddedeb-88e8-438e-904c-ddcbfb0abf5b is stale, updating it" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:53Z" level=info msg="Updating IPSet myapp-crowdsec-ipset-a-IPV4-ban-01ddedeb-88e8-438e-904c-ddcbfb0abf5b" set=myapp-crowdsec-ipset-a-IPV4-ban-01ddedeb-88e8-438e-904c-ddcbfb0abf5b
time="2024-11-03T16:49:53Z" level=info msg="Creating IPSet myapp-crowdsec-ipset-a-IPV4-ban-01ddedeb-88e8-438e-904c-ddcbfb0abf5b" set=myapp-crowdsec-ipset-a-IPV4-ban-01ddedeb-88e8-438e-904c-ddcbfb0abf5b
time="2024-11-03T16:49:54Z" level=info msg="set myapp-crowdsec-ipset-a-IPV4-ban-262d7d94-8f87-4f33-880c-a16a8645bc7e is stale, updating it" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:54Z" level=info msg="Updating IPSet myapp-crowdsec-ipset-a-IPV4-ban-262d7d94-8f87-4f33-880c-a16a8645bc7e" set=myapp-crowdsec-ipset-a-IPV4-ban-262d7d94-8f87-4f33-880c-a16a8645bc7e
time="2024-11-03T16:49:54Z" level=info msg="Creating IPSet myapp-crowdsec-ipset-a-IPV4-ban-262d7d94-8f87-4f33-880c-a16a8645bc7e" set=myapp-crowdsec-ipset-a-IPV4-ban-262d7d94-8f87-4f33-880c-a16a8645bc7e
time="2024-11-03T16:49:55Z" level=info msg="set myapp-crowdsec-ipset-a-IPV4-ban-3aab357a-5fc4-4fca-97c8-4afcbdaac5a4 is stale, updating it" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:55Z" level=info msg="Updating IPSet myapp-crowdsec-ipset-a-IPV4-ban-3aab357a-5fc4-4fca-97c8-4afcbdaac5a4" set=myapp-crowdsec-ipset-a-IPV4-ban-3aab357a-5fc4-4fca-97c8-4afcbdaac5a4
time="2024-11-03T16:49:55Z" level=info msg="Creating IPSet myapp-crowdsec-ipset-a-IPV4-ban-3aab357a-5fc4-4fca-97c8-4afcbdaac5a4" set=myapp-crowdsec-ipset-a-IPV4-ban-3aab357a-5fc4-4fca-97c8-4afcbdaac5a4
time="2024-11-03T16:49:57Z" level=info msg="set myapp-crowdsec-ipset-a-IPV6-ban-2870c4dd-dd7c-4546-af95-952f65294eb8 is stale, updating it" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
time="2024-11-03T16:49:57Z" level=info msg="Updating IPSet myapp-crowdsec-ipset-a-IPV6-ban-2870c4dd-dd7c-4546-af95-952f65294eb8" set=myapp-crowdsec-ipset-a-IPV6-ban-2870c4dd-dd7c-4546-af95-952f65294eb8
time="2024-11-03T16:49:57Z" level=info msg="Creating IPSet myapp-crowdsec-ipset-a-IPV6-ban-2870c4dd-dd7c-4546-af95-952f65294eb8" set=myapp-crowdsec-ipset-a-IPV6-ban-2870c4dd-dd7c-4546-af95-952f65294eb8
time="2024-11-03T16:49:58Z" level=warning msg="Dependencies of rule group crowdsec-waf-bouncer-ip-set-myapp not ready yet, retrying in 2 seconds"
time="2024-11-03T16:50:00Z" level=info msg="Polling decisions"
time="2024-11-03T16:50:09Z" level=info msg="Polling decisions"
```

> How many ACLs is the bouncer configured to manage ?

Just one, with the name `crowdsec-staging` which configuration is the one shared in the Issue's original message

This is the IAM policy regarding the WAF for every EC2 instance in my case:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "wafv2:UpdateWebACL",
                "wafv2:UpdateRuleGroup",
                "wafv2:UpdateIPSet",
                "wafv2:TagResource",
                "wafv2:GetWebACL",
                "wafv2:GetRuleGroup",
                "wafv2:GetIPSet",
                "wafv2:DeleteRuleGroup",
                "wafv2:DeleteIPSet",
                "wafv2:CreateRuleGroup",
                "wafv2:CreateIPSet"
            ],
            "Effect": "Allow",
            "Resource": [
                "arn:aws:wafv2:*:*:regional/webacl/*/*",
                "arn:aws:wafv2:*:*:regional/rulegroup/*/*",
                "arn:aws:wafv2:*:*:regional/managedruleset/*/*",
                "arn:aws:wafv2:*:*:*/ipset/*/*"
            ],
            "Sid": "WAF1"
        },
        {
            "Action": [
                "wafv2:ListWebACLs",
                "wafv2:ListRuleGroups",
                "wafv2:ListIPSets"
            ],
            "Effect": "Allow",
            "Resource": "*",
            "Sid": "WAF2"
        }
    ]
}
```

**Comment by blotus** [comment]
Have you manually cleaned up the ipsets before starting the bouncer ? (ie, is the bouncer able to create new ipsets ?).
If the bouncer cannot create a new set, it should return an error.
If it's not detected for some reason, it could probably desynchronize the internal state compared to what actually exists in AWS.

Also, based on your logs, it seems that the bouncer does not even receive the `SIGTERM` signal (a `terminating bouncer process` should appear in your logs the moment you issue the `stop` command), which could also explain the sets not being deleted. 

Can you share your `docker-compose.yml` file? Have you specified a custom `stop_signal` in it? The bouncer expects `SIGTERM` or `SIGINT` on stop. If you send any other signal, it will not be caught and will kill the process instantly without a change for cleanup.

When the container is running, you can also run `docker inspect <BOUNCER_CONTAINER_NAME> | jq '.[].Config.StopSignal'` to see if there's a custom signal defined for the container.

**Comment by lgg42** [comment]
> Have you manually cleaned up the ipsets before starting the bouncer?

Not before I hit the 100 IPsets limit. After that I started clearing IPsets manually. 

> ie, is the bouncer able to create new ipsets ?

Yes, in the previous message before stopping the bouncer container I had 21 IPsets belonging to myapp. After starting again the bouncer container I had 27 IPsets. These new IPsets are updated as new decisions are triggered or IP unban commands are executed.

>  Have you specified a custom `stop_signal` in it? The bouncer expects `SIGTERM` or `SIGINT` on stop.
AFAIK, nope, here's the compose file regarding the bouncer and crowdsec

```yaml
services:

  # crowdsec: It will be fed nginx's logs and will signal the firewall bouncer
  crowdsec:
    image: crowdsecurity/crowdsec:v1.6.2
    restart: always
    environment:
      COLLECTIONS: "crowdsecurity/nginx"
      GID: "${GID-1000}"
    depends_on:
      - 'reverse-proxy'
    volumes:
      - ./crowdsec/acquis.yaml:/etc/crowdsec/acquis.yaml
      - ./crowdsec/parsers/s02-enrich/01-node-modules.yaml:/etc/crowdsec/parsers/s02-enrich/01-node-modules.yaml
      - ./logs/nginx:/var/log/nginx
      - crowdsec-db:/var/lib/crowdsec/data/
      - crowdsec-config:/etc/crowdsec/
    network_mode: "host"
    profiles: [crowdsec]

  # crowdsec-awf-waf-bouncer: Automatically adds rules to an AWS WAF ACL and manages IPSets content to apply decisions taken by crowdsec
  crowdsec-awf-waf-bouncer:
    image: crowdsecurity/aws-waf-bouncer:v0.1.7
    restart: always
    environment:
      BOUNCER_CONFIG_FILE: /crowdsec-aws-waf-bouncer.yaml
    depends_on:
      - 'crowdsec'
    volumes:
      - ./crowdsec/cs-aws-waf-bouncer-staging.yaml:/crowdsec-aws-waf-bouncer.yaml # AWS WAS Bouncer configuration
    network_mode: "host"
    profiles: [crowdsec]
```

> `docker inspect <BOUNCER_CONTAINER_NAME> | jq '.[].Config.StopSignal'`

The output of this comand is:
```shell
null
```

Thanks again for your time :pray: if there are any extra tests I could carry out I'm all :eyes:

**Comment by blotus** [comment]
I've managed to reproduce the issue, it seems specific to docker, I'm still not sure what is happening but I can confirm there's a bug.

**Comment by blotus** [comment]
Hello @lgg42,

I've just released v0.1.8, which fixes the issue.
Please update your docker image and let me know if everything looks good to you.

**Comment by lgg42** [comment]
Woah, this was fast and sweet! :rainbow: 

I can confirm from my side that with v0.1.8 the bouncer is deleting the IPsets upon stop signal. These are the logs after the stop command is executed:

```
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:45Z" level=info msg="deleting set myapp-crowdsec-ipset-a-IPV4-ban-0c84faba-59e6-46c4-b596-f8becc16ba4a" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:45Z" level=info msg="Deleting IPSet myapp-crowdsec-ipset-a-IPV4-ban-0c84faba-59e6-46c4-b596-f8becc16ba4a" set=myapp-crowdsec-ipset-a-IPV4-ban-0c84faba-59e6-46c4-b596-f8becc16ba4a
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:46Z" level=info msg="deleting set myapp-crowdsec-ipset-a-IPV4-ban-c34dc677-393c-4d81-b8ae-e8a083ee58da" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:46Z" level=info msg="Deleting IPSet myapp-crowdsec-ipset-a-IPV4-ban-c34dc677-393c-4d81-b8ae-e8a083ee58da" set=myapp-crowdsec-ipset-a-IPV4-ban-c34dc677-393c-4d81-b8ae-e8a083ee58da
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:47Z" level=info msg="deleting set myapp-crowdsec-ipset-a-IPV6-ban-98b4c0b6-e840-4995-9e02-644a65cc0bd3" acl=crowdsec-staging component=ipset_manager region=us-east-1 scope=REGIONAL
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:47Z" level=info msg="Deleting IPSet myapp-crowdsec-ipset-a-IPV6-ban-98b4c0b6-e840-4995-9e02-644a65cc0bd3" set=myapp-crowdsec-ipset-a-IPV6-ban-98b4c0b6-e840-4995-9e02-644a65cc0bd3
myapp-crowdsec-awf-waf-bouncer-1  | time="2024-11-04T09:47:47Z" level=fatal msg="process terminated with error: received SIGTERM"
myapp-crowdsec-awf-waf-bouncer-1 exited with code 1
myapp-crowdsec-awf-waf-bouncer-1 exited with code 0
```

After every stop/restart there are no IPsets remaining in the WAF console. Everything is clean :raised_hands: 

Thank you for your work @blotus :bow:
